machida/lib/wallaroo/experimental/__init__.py

```diff
@@ -100,6 +125,325 @@ def write(self, message, event_time=0, key=None):
+                                   point_of_ref = point_of_ref,
+                                   is_open = False)
+            # set handshake_complete
+            self.handshake_complete = True
+            # set terminator to 4
+            self.set_terminator(4)
+            # set socket to nonblocking
+            self._conn.setblocking(0)
+            # initialize the asynchat handler
+            asynchat.async_chat.__init__(self, sock=self._conn)
+
+    def _handle_notify_ack(self, msg):
+        self.update_stream(msg.stream_id,
+                           point_of_ref = msg.point_of_ref,
+                           is_open = msg.notify_success)
+
+    def _handle_ack(self, msg):
+        self.credits += msg.credits
+        for (stream_id, point_of_ref) in msg.acks:
+            self.update_stream(stream_id,
+                               point_of_ref = point_of_ref)
+
+    def update_stream(self, stream_id, stream_name=None,
+                      point_of_ref=None, is_open=None):
+        old = self._streams.get(stream_id, None)
+        if old is None:
+            new = Stream(stream_id, stream_name, point_of_ref, is_open)
+        else:
+            if stream_name is not None:
+                if old.name != steam_name:
+                    # stream collision... throw error and close connection
+                    raise ConnectorError("Got wrong stream name for "
+                                         "stream. Expected {} but got {}."
+                                         .format(old.name, stream_name))
+            new = Stream(stream_id,
+                         old.name if old is not None else stream_name,
+                         (point_of_ref if point_of_ref is not None else
+                          old.point_of_ref),
+                         is_open if is_open is not None else old.is_open)
+        # update the stream information
+        self._streams[stream_id] = new
+        # Call optional user handler with updated stream
+        self.stream_updated(new)
+
+    ##########################
+    # Outoing communications #
+    ##########################
+
+    def initiate_handshake(self, hello):
+        conn = socket.socket()
+        conn.connect( (self._host, self._port) )
+        self._conn = conn
+
+        self.in_handshake = True
+        self.version = hello.version
+        self.cookie = hello.cookie
+        self.program_name = hello.program_name
+        self.instance_name = hello.instance_name
+        self._conn.setblocking(1) # Set socket to blocking mode
+        self._conn.sendall(cwm.Frame.encode(hello))
+        header_bytes = self._conn.recv(4)
+        frame_size = struct.unpack('>I', header_bytes)[0]
+        frame = self._conn.recv(frame_size)
+        try:
+            self._handle_frame(frame)
+        except Exception as err:
+            # close the connection and raise the error
+            self._conn.close()
+            raise err
+
+    def write(self, msg):
```

Comment:
Need to be careful about what happens in this synchronous call when an async background handler changes the state we're comparing against here.

Reply:
I wrote more in some other comments but the more I think over it, I think we should keep the event-based API async and use readiness as a trigger to call something on the class rather than the other way around. We can then use a bounded queue and some state variables to interact between a threaded+blocking world and the event loop+reactor by using a generic state machine tied to the queue as the class we use to implement all of the callbacks. This machine could also include concerns that are API but not protocol related like automatic reconnection as well as retry limits or back-off, and so on.